From 2ab7339d20800944054cd7c794dfebf5ba4fe1e0 Mon Sep 17 00:00:00 2001
Message-Id: <2ab7339d20800944054cd7c794dfebf5ba4fe1e0.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Tue, 3 Jun 2014 21:57:55 +0530
Subject: [PATCH 083/312] FOR_UPSTREAM [VPG]: drm/i915: WAR for memory
 bandwidth limited devices

Following changes are ported:

	WAR for memory bandwidth limited devices
	(OriginalChangeID - Ia134a1bce587532d7a83d4cf5f4e1e5dd0d5e49a)
	Description: When there is memory bandwidth limitation, there
	is corruption while playing dual display, with HDMI modes
	above 720P resolution. This is a work around patch to prune modes above 1280x800
	resolution, while parsing HDMI EDID.

Issue: GMIN-1443
Change-Id: I8eab055bbc7cdea14255fc20ee418c257fb22080
Signed-off-by: Shashank Sharma <shashank.sharma@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h    |    1 +
 drivers/gpu/drm/i915/i915_params.c |    4 ++++
 drivers/gpu/drm/i915/intel_hdmi.c  |   26 ++++++++++++++++++++++++++
 3 files changed, 31 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 3b4489a..3d8f0ab 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2088,6 +2088,7 @@ extern void i915_master_destroy(struct drm_device *dev, struct drm_master *maste
 
 /* i915_params.c */
 struct i915_params {
+	int limitbw;
 	int modeset;
 	int panel_ignore_lid;
 	unsigned int powersave;
diff --git a/drivers/gpu/drm/i915/i915_params.c b/drivers/gpu/drm/i915/i915_params.c
index 2ebb5fc..1d5eb7a 100644
--- a/drivers/gpu/drm/i915/i915_params.c
+++ b/drivers/gpu/drm/i915/i915_params.c
@@ -25,6 +25,7 @@
 #include "i915_drv.h"
 
 struct i915_params i915 __read_mostly = {
+	.limitbw = 0,
 	.modeset = -1,
 	.panel_ignore_lid = 1,
 	.powersave = 1,
@@ -50,6 +51,9 @@ struct i915_params i915 __read_mostly = {
 	.disable_vtd_wa = 0,
 	.drrs_interval = 2000,
 };
+module_param_named(limitbw, i915.limitbw, int, 0400);
+MODULE_PARM_DESC(limitbw,
+	"Override lid status (0=not limited [default], 1=limited bandwidth, ");
 
 module_param_named(modeset, i915.modeset, int, 0400);
 MODULE_PARM_DESC(modeset,
diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index 5782b6f..64e37c6 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -37,6 +37,21 @@
 #include <drm/i915_drm.h>
 #include "i915_drv.h"
 
+#define HDMI_UEVENT_MAX_LENGTH 20
+
+#define LIMIT_BW_MAX_HDISPLAY	1280
+#define LIMIT_BW_MAX_VDISPLAY	800
+
+/* CEA Mode 4 - 1280x720@60Hz */
+struct drm_display_mode hdmi_fallback_mode = {
+	DRM_MODE("1280x720", DRM_MODE_TYPE_DRIVER,
+	74250,
+	1280, 1390, 1430, 1650, 0,
+	720, 725, 730, 750, 0,
+	DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_PVSYNC),
+	.vrefresh = 60
+};
+
 static struct drm_device *intel_hdmi_to_dev(struct intel_hdmi *intel_hdmi)
 {
 	return hdmi_to_dig_port(intel_hdmi)->base.base.dev;
@@ -911,12 +926,23 @@ static enum drm_mode_status
 intel_hdmi_mode_valid(struct drm_connector *connector,
 		      struct drm_display_mode *mode)
 {
+	struct drm_device *dev = connector->dev;
 	if (mode->clock > hdmi_portclock_limit(intel_attached_hdmi(connector),
 					       true))
 		return MODE_CLOCK_HIGH;
 	if (mode->clock < 20000)
 		return MODE_CLOCK_LOW;
 
+	/* WAR
+	 * For limitation in memory chanel bandwidth.
+	 */
+	if ((i915.limitbw) && (IS_VALLEYVIEW(dev))) {
+		if (mode->vdisplay > LIMIT_BW_MAX_VDISPLAY)
+			return MODE_BAD_VVALUE;
+		if (mode->hdisplay > LIMIT_BW_MAX_HDISPLAY)
+			return MODE_BAD_HVALUE;
+	}
+
 	if (mode->flags & DRM_MODE_FLAG_DBLSCAN)
 		return MODE_NO_DBLESCAN;
 
-- 
1.7.9.5

